### PR TITLE
Set allocation bit on allocation fast path

### PR DESCRIPTION
**Pull request by YiluoWei**

### CODE REVIEW

File: openjdk/mmtkBarrierSet.hpp [hunk 1]
```diff
@@ -37,6 +37,9 @@
 #include "mmtkBarrierSetAssembler_x86.hpp"
 
 #define MMTK_ENABLE_ALLOCATION_FASTPATH true
+#define MMTK_ENABLE_GLOBAL_ALLOC_BIT false
+// hard coded for now
+const intptr_t ALLOC_BIT_BASE_ADDRESS = 0x0000060000000000;
```
Comment (qinsoon) [inline openjdk/mmtkBarrierSet.hpp:42]: If you want to get the address, you can do something like this: https://github.com/mmtk/mmtk-openjdk/blob/648bb27f4bb5c84a83f74fda1d699514548cf663/openjdk/mmtk.h#L15

File: openjdk/mmtkBarrierSetC2.cpp [hunk 2]
```diff
@@ -281,6 +281,58 @@ void MMTkBarrierSetC2::expand_allocate(
     fast_oop_ctrl = needgc_false; // No contention, so this is the fast path
     fast_oop_rawmem = store_eden_top;
 
+#if MMTK_ENABLE_GLOBAL_ALLOC_BIT
+    Node *obj_addr = new CastP2XNode(fast_oop_ctrl, fast_oop);
```
Comment (qinsoon) [inline openjdk/mmtkBarrierSetC2.cpp:285]: Need some comments for this part. Maybe put the expected assembly as comments.

**Review by qinsoon — COMMENTED** [review]
We need a feature for `mmtk_openjdk` to enable `global_alloc_bit` for `mmtk_core`. You can follow what the currently build file does and have an env var for it. If the env var is set, enable the feature and enable the alloc bit in fast path for the binding. We should also test alloc bit in `ci-test.sh` (make sure your assertions about alloc bit in mmtk-core is tested).

Once those are done, can you also run some experiments to see if the overhead for the alloc bit is as expected?

File: openjdk/mmtkBarrierSet.hpp (hunk 1 above)
Comment (k-sareen) [reply openjdk/mmtkBarrierSet.hpp:42]: See this PR: https://github.com/mmtk/mmtk-openjdk/pull/86

**Comment by YiluoWei** [comment]
Could not pass some extreme assertions, because the allocation bit was set in the fast path, thus by-passed the side-metadata module.  It actually showed the allocation bit was set when tracing the object, which was expected.

```
yiluow@rat:~/mmtk-dev/debug/mmtk-openjdk/repos/openjdk$ export MMTK_PLAN=GenCopy; RUST_BACKTRACE=1 ./build/linux-x86_64-normal-server-$DEBUG_LEVEL/jdk/bin/java -server -XX:MetaspaceSize=500M -XX:+UseThirdPartyHeap -Xms$HEAP_SIZE -Xmx$HEAP_SIZE  -XX:-UseBiasedLocking -XX:-TieredCompilation -Xcomp -jar $DACAPO  -n 1 luindex
[2021-08-24T10:16:53Z INFO  mmtk::memory_manager] Initialized MMTk with GenCopy
[2021-08-24T10:16:53Z WARN  mmtk::memory_manager] The feature 'extreme_assertions' is enabled. MMTk will run expensive run-time checks. Slow performance should be expected.
[2021-08-24T10:17:06Z INFO  mmtk::plan::global] User triggering collection
thread '<unnamed>' panicked at 'verify_load(SideMetadataSpec { **is_global: true **offset: 0x60000000000 **log_num_of_bits: 0x0 **log_min_obj_size: 0x3 }, 0x20000b99d60) -> Expected (0x0) but found (0x1)', /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/util/metadata/side_metadata/sanity.rs:426:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/ae90dcf0207c57c3034f00b07048d63f8b2363c8/library/std/src/panicking.rs:517:5
   1: std::panicking::begin_panic_fmt
             at /rustc/ae90dcf0207c57c3034f00b07048d63f8b2363c8/library/std/src/panicking.rs:460:5
   2: mmtk::util::metadata::side_metadata::sanity::verify_load
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/util/metadata/side_metadata/sanity.rs:426:13
   3: mmtk::util::metadata::side_metadata::global::load_atomic
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/util/metadata/side_metadata/global.rs:360:5
   4: mmtk::util::alloc_bit::is_alloced_object
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/util/alloc_bit.rs:84:5
   5: mmtk::util::alloc_bit::is_alloced
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/util/alloc_bit.rs:80:5
   6: mmtk::policy::copyspace::CopySpace<VM>::trace_object
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/policy/copyspace.rs:189:13
   7: <mmtk::plan::gencopy::gc_work::GenCopyNurseryProcessEdges<VM> as mmtk::scheduler::gc_work::ProcessEdgesWork>::trace_object
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/plan/gencopy/gc_work.rs:106:20
   8: <mmtk::plan::gencopy::gc_work::GenCopyNurseryProcessEdges<VM> as mmtk::scheduler::gc_work::ProcessEdgesWork>::process_edge
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/plan/gencopy/gc_work.rs:132:26
   9: mmtk::scheduler::gc_work::ProcessEdgesWork::process_edges
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/scheduler/gc_work.rs:460:13
  10: mmtk::scheduler::gc_work::<impl mmtk::scheduler::work::GCWork<<E as mmtk::scheduler::gc_work::ProcessEdgesWork>::VM> for E>::do_work
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/scheduler/gc_work.rs:470:9
  11: mmtk::scheduler::work::GCWork::do_work_with_stat
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/scheduler/work.rs:20:9
  12: mmtk::scheduler::worker::GCWorker<VM>::run
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/scheduler/worker.rs:158:13
  13: mmtk::memory_manager::start_worker
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/repos/mmtk-core/src/memory_manager.rs:177:5
  14: start_worker
             at /home/yiluow/mmtk-dev/debug/mmtk-openjdk/mmtk/src/api.rs:123:5
  15: _ZN6Thread8call_runEv
  16: _ZL19thread_native_entryP6Thread
  17: start_thread
  18: __clone
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5
Aborted (core dumped)
```

File: .github/scripts/ci-test-global-alloc-bit.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+# Build
+sh configure --disable-warnings-as-errors --with-debug-level=$DEBUG_LEVEL
+make CONF=linux-x86_64-normal-server-$DEBUG_LEVEL THIRD_PARTY_HEAP=$PWD/../../openjdk
+
+# --- SemiSpace ---
+export MMTK_PLAN=SemiSpace
+
+# Test - the benchmarks that are commented out do not work yet
+# Note: the command line options are necessary for now to ensure the benchmarks work. We may later change the options if we do not have these many constraints.
+build/linux-x86_64-normal-server-$DEBUG_LEVEL/jdk/bin/java -XX:+UseThirdPartyHeap -server -XX:MetaspaceSize=100M -Xms500M -Xmx500M -XX:-TieredCompilation -Xcomp -jar benchmarks/dacapo-2006-10-MR2.jar antlr
```
Comment (caizixian) [inline .github/scripts/ci-test-global-alloc-bit.sh:24]: We can't just keep duplicating the scripts ;-)

**Review by caizixian — CHANGES_REQUESTED** [review]

File: openjdk/mmtkBarrierSet.hpp [hunk 4]
```diff
@@ -38,6 +38,8 @@
 
 #define MMTK_ENABLE_ALLOCATION_FASTPATH true
 
+const intptr_t ALLOC_BIT_BASE_ADDRESS = GLOBAL_SIDE_METADATA_BASE_ADDRESS;
```
Comment (qinsoon) [inline openjdk/mmtkBarrierSet.hpp:41]: You are assuming the first global side metadata spec is always the global alloc bit. It is true for now, but there is no way to guarantee that this won't be changed. Like in your alloc bit PR, you put the alloc bit metadata to the first. Others may have this kind of change as well. You need the address for the alloc bit.

**Review by qinsoon — APPROVED** [review]
LGTM

**Comment by qinsoon** [comment]
I think this PR is good now. If you plan to provide any comment, please do so as soon as you can. Otherwise I would get this merged before the v0.7 release.

**Pull request merged**